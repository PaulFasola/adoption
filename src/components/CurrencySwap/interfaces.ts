import { IProtocol } from '../ProtocolSelector/interfaces';

export interface IProtocolArrayPipe {
  input: IProtocol[];
  output: IProtocol[];
}

export interface IProtocolPipe {
  input?: IProtocol;
  output?: IProtocol;
}

export interface IProps {
  locked?: boolean;
  noShadow?: boolean;
  protocols: Partial<IProtocolArrayPipe>;
  maxFractionDigits?: number;

  onSubmit?: () => void;
}

export interface ISwapValues {
  input?: number;
  output?: number;
}
