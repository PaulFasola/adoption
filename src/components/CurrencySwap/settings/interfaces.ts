import { IProps as IInputProps } from '../../common/Input/interfaces';
import { IProps } from '../interfaces';

export interface ISettingsProps extends Pick<IProps, 'settings'> {
  settings: ISettings;
  visible?: boolean;
  renderable?: boolean;
}

export interface ISettings {
  [key: string]: ISetting;
}

export type SettingType = 'boolean' | 'string' | 'number' | 'radio';

export interface ISetting {
  visible: boolean;
  type: SettingType;
  label: string;
  hint?: string;
  customInput?: IInputProps;
  value?: any;
}
