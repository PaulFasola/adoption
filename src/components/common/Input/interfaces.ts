export interface InputProps {
  id?: string;
  pattern?: string;
  className?: string;
  maxLength?: number;
  minLength?: number;
  max?: number;
  min?: number;
  placeholder?: string;
  title?: string;
  value?: string | number;
  label?: string;
  autoComplete?: 'on' | 'off';
  autoFill?: 'on' | 'off';
  style?: React.CSSProperties;

  onValueChange?: (value: string) => void;
}

export interface IProps extends InputProps {
  type: 'text' | 'number' | 'decimal';
  maximumFractionDigits?: number;
  style?: React.CSSProperties;
}
