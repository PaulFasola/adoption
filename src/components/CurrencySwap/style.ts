import styled from 'styled-components';
import _ from '../../providers/theme/styleFetcher';
import { ITheme } from '../../providers/theme/ITheme';
import { Input } from '../common/Input';

export interface IStyleProps {
  theme: ITheme;
  noShadow?: boolean;
}

interface SP extends IStyleProps {}

export const Container = styled.div`
  position: relative;
  padding: 15px;
  min-height: 190px;
  max-height: 800px;
  min-width: 250px;
  max-width: 350px;
  background-color: ${(p) => _(p.theme, 'primary', 'backgroundColor')};
  box-shadow: ${(p) => _(p.theme, 'primary', 'boxShadow')};

  ${(p: SP) =>
    p.noShadow &&
    `
    box-shadow: none;
  `}
`;

export const InputWrapper = styled.div`
  position: relative;
  border: 1px solid ${(p) => _(p.theme, 'primary', 'borderColor')};
  background-clip: padding-box;

  span:first-of-type {
    display: inline-block;
    margin: 5px 5px 0 5px;
    color: ${(p) => _(p.theme, 'primary', 'color')};
  }

  & > div {
    display: flex;
    align-items: center;

    /* ProtocolSelector wrapper */
    div {
      margin-right: 0.2em;
    }
  }
`;

export const SwapInput = styled(Input)`
  border: none;
  width: calc(100% - 10px);
  font-size: 1.2em;
`;

export const SwapButton = styled.button`
  display: block;
  height: 50px;
  width: 50px;
  margin: auto;
  padding: 0;
  outline: 0;
  border: none;
  background: transparent;
  cursor: pointer;

  &:hover path,
  &:focus path {
    transition: fill 0.15s ease-in;
    fill: ${(p) => _(p.theme, 'disabled', 'color')}!important;
  }

  path {
    fill: ${(p) => _(p.theme, 'primary', 'color')}!important;
  }

  > div {
    display: block;
    width: 35px;
    margin: auto;
    overflow: visible;
  }
`;

export const SubmitButton = styled.button`
  position: relative;
  font-size: 1.2em;
  background-color: ${(p) => _(p.theme, 'defaultButton', 'backgroundColor')};
  border: 0;
  color: ${(p) => _(p.theme, 'secondary', 'color')};
  width: 100%;
  margin: 1rem auto 0 auto;
  padding: 15px;
  outline: none;
  cursor: pointer;

  &:disabled {
    color: ${(p) => _(p.theme, 'disabled', 'color')};
    background-color: ${(p) => _(p.theme, 'disabled', 'backgroundColor')};
    cursor: auto;
  }

  &:hover,
  &:focus {
    &:not(:disabled) {
      background-color: ${(p) => _(p.theme, 'defaultButton', 'hoverBgColor')};
    }
  }
`;

export const Overview = styled.span`
  position: absolute;
  right: 10px;
  top: 8px;
  font-size: 0.7em;
  text-transform: uppercase;
  color: ${(p) => _(p.theme, 'primary', 'color')};
`;
