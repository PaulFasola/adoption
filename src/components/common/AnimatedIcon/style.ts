import styled from 'styled-components';
import _ from '../../../providers/theme/styleFetcher';
import { stroke } from '../animations';

export const Circle = styled.circle`
  stroke-dasharray: 166;
  stroke-dashoffset: 166;
  stroke-width: 2;
  stroke-miterlimit: 10;
  stroke: #4bb71b;
  fill: ${(p) => _(p.theme, 'primary', 'backgroundColor')};
  animation: ${stroke} 0.6s cubic-bezier(0.65, 0, 0.45, 1) forwards;
`;

export const Container = styled.svg`
  width: 100px;
  height: 100px;
  display: block;
  position: relative;
  margin: 0 auto;
`;

export const CheckMark = styled.path`
  stroke-width: 2;
  stroke-miterlimit: 10;
  box-shadow: inset 0px 0px 0px #4bb71b;
  border-radius: 50%;
  stroke: #4bb71b;
  transform-origin: 50% 50%;
  stroke-dasharray: 48;
  stroke-dashoffset: 48;
  animation: ${stroke} 0.3s cubic-bezier(0.65, 0, 0.45, 1) 0.8s forwards;
`;

export const Times = styled.g`
  path {
    stroke-width: 2;
    stroke: #4bb71b;
    stroke-dasharray: 48;
    stroke-dashoffset: 48;
    transform-origin: 50% 50% 0;
    animation: 0.3s ease 0.8s normal forwards 1 running ${stroke};
  }
`;

export const IconContainer = styled.div`
  display: inline-block;
  vertical-align: middle;
`;
