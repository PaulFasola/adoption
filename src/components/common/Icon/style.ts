import styled from 'styled-components';
import _ from '../../../providers/theme/styleFetcher';

export const Link = styled.a`
  color: rgba(0, 0, 0, 0.8);
  text-decoration: none;
`;

export const IconContainer = styled.div`
  margin: 5px 3px 3px 3px;
  display: inline-block;
  vertical-align: middle;

  svg .fillable {
    fill: ${(p) => _(p.theme, 'primary', 'color')};
  }
`;
