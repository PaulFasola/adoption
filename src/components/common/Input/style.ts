import styled from 'styled-components';
import _ from '../../../providers/theme/styleFetcher';

export interface IStyleProps {
  disabled?: boolean;
  floating?: boolean;
  rounded?: boolean;
}

interface SP extends IStyleProps {}

export const Outline = styled.div`
  width: 100%;
`;

export const Input = styled.input`
  background: transparent;
  color: ${({ theme }) => _(theme, 'primary', 'color')};
  outline: none;
  padding: 10px;
  box-shadow: none;
  text-overflow: ellipsis;

  ${(props: SP) =>
    props.rounded &&
    `
		border-radius: 15px;
	`}

  &: focus {
    outline: none;
    box-shadow: inset 0 -0.25em 0.5em lighten(#000, 50%);
    border-width: 2px;
  }

  &:-moz-ui-invalid {
    box-shadow: none !important;
  }
`;

export const Label = styled.label`
  position: absolute;
  top: 25px;
  left: 25px;
  padding: 0 2px;
  line-height: 20px;
  font-size: 14px;
  letter-spacing: 0.125em;
  color: ${(p) => _(p.theme, 'primary', 'color')};
  background: ${(p) => _(p.theme, 'primary', 'backgroundColor')};
  pointer-events: none;
  transition: transform 120ms;
  transform-origin: left center;
  transform: translateY(0) scale(1);

  ${(props: SP) => props.floating && 'transform: translateY(-19px) scale(0.9);'}
`;
