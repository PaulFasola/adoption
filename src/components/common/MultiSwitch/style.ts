import styled from 'styled-components';
import { ITheme } from '../../../providers/theme/ITheme';
import _ from '../../../providers/theme/styleFetcher';
import { Input } from '../Input';
import { centerContent } from '../style';

interface IStyleProps {
  theme: ITheme;
  hasCustomValue?: boolean;
  isCustomFieldFocused?: boolean;
}

interface SP extends IStyleProps {}

const getFieldWidth = (hasCustomValue?: boolean) => {
  // TODO: handle size for 3 & 4 radios
  return hasCustomValue ? 65 : 82;
};

export const Slider = styled.span`
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  background-color: ${(p) => _(p.theme, 'disabled', 'color')};
`;

const getInputTransitions = (hasCustomValue?: boolean): string => {
  const fieldWidth = getFieldWidth(hasCustomValue);
  let style = '';

  style += `
    div ~ ${Slider} {
      transition: left .2s ease-in-out;
      left: ${fieldWidth * 5 - fieldWidth + 1}px;
    }
  `;

  for (let i = 1; i < 5; i++) {
    style += `
      input[type='radio']:nth-of-type(${i}):checked ~ ${Slider} {
        transition: left .2s ease-in-out;
        left: ${fieldWidth * i - fieldWidth}px;
      }
    `;
  }

  return style;
};

export const Label = styled.label`
  ${centerContent}
  z-index: 1;
  position: relative;
  height: 100%;
  float: left;
  cursor: pointer;
  color: ${(p) => _(p.theme, 'primary', 'color')};
  transition: all 500ms ease-in-out;

  &:focus {
    outline: none;
  }

  span {
    user-select: none;
  }
`;

export const CustomValue = styled(Input)`
  position: relative;
  z-index: 1;
  outline: none;
  border: none;
  box-sizing: border-box;
  font-size: 1em;
  overflow: hiddenl
  width: 65px;

  &:placeholder-shown {
    color: ${(p) => _(p.theme, 'disabled', 'color')};
  }

  &:focus {
    color: ${(p) => _(p.theme, 'primary', 'color')};
  }

  ::selection {
    color: ${(p) => _(p.theme, 'primary', 'backgroundColor')};
    background: ${(p) => _(p.theme, 'primary', 'color')};
  }
`;

export const Container = styled.div`
  overflow: hidden;
  position: relative;
  flex-wrap: nowrap;
  box-sizing: border-box;
  display: inline-block;
  padding: 0;
  width: 100%;
  height: 2rem;
  background-clip: padding-box;
  border: 2px solid ${(p) => _(p.theme, 'primary', 'borderColor')};
  transition: left 1.8s ease-in-out;
  ${centerContent}

  input[type='radio'] {
    position: absolute;
    visibility: hidden;
  }

  div {
    display: inline-block;
    & > input {
      padding: 0 15px 0 8px;
      width: 95% !important;
    }
  }

  span.suffix {
    position: absolute;
    z-index: 1;
    right: 0;
    padding-right: 5px;
  }

  ${({ hasCustomValue }: SP) => `
    ${Label}, ${Slider}, ${CustomValue}, div {
      width: ${getFieldWidth(hasCustomValue)}px;
    }
  `}

  ${({ hasCustomValue }: SP) => getInputTransitions(hasCustomValue)}
`;
