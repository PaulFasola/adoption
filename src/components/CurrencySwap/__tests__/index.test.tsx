import React from 'react';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import { activeProtocols } from '../../ProtocolSelector/__tests__/mocks';
import { configure, mount, ReactWrapper, shallow } from 'enzyme';
import { CurrencySwap } from '..';
import { Overview, SwapButton } from '../style';
import { ISwapSubmission, ProtocolEnd } from '../interfaces';
import { IProtocol } from '../../ProtocolSelector/interfaces';

const SUBMIT_BUTTON_SELECTOR = 'button[aria-label="Submit"]';

expect.extend(toHaveNoViolations);
configure({ adapter: new Adapter() });

const selectProtocol = (
  wrapper: ReactWrapper,
  select: ProtocolEnd,
  protoPos: number
): ReactWrapper => {
  if (protoPos < 0) throw "Protocol DOM poisiton can't be < 0.";

  const selectPos = select === 'input' ? 0 : 1;
  const button = wrapper
    .find('button')
    .not('[aria-label="Swap values"]')
    .at(selectPos)
    .simulate('click');

  wrapper
    .find('div[data-test="dropdown"]')
    .at(selectPos)
    .find('div > div')
    .at(protoPos)
    .simulate('click');

  return button;
};

const setInputValue = (
  wrapper: ReactWrapper,
  inputEnd: ProtocolEnd,
  amount: string | number
): ReactWrapper => {
  const input = wrapper.find('input').at(inputEnd === 'input' ? 0 : 1);
  input.simulate('change', { target: { value: amount.toString() } });

  return input;
};

describe('CurrencySwap component', () => {
  it('should render', () => {
    shallow(
      <CurrencySwap
        protocols={{
          input: activeProtocols,
        }}
      />
    );
  });

  /**
   * VALIDATION LOGIC
   */
  it('should respect maxFractionDigit prop value', () => {
    shallow(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
      />
    );
  });

  it('should prevent submission when', () => {
    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
      />
    );

    expect(wrapper.find(SUBMIT_BUTTON_SELECTOR).getDOMNode()).toBeDisabled();
  });

  it('should prevent submission when', () => {
    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
      />
    );

    expect(wrapper.find(SUBMIT_BUTTON_SELECTOR).getDOMNode()).toBeDisabled();
  });

  it('should prevent submission when', () => {
    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
      />
    );

    expect(wrapper.find(SUBMIT_BUTTON_SELECTOR).getDOMNode()).toBeDisabled();
  });

  it('should render balance when an input protocol is selected', () => {
    const testValue = 42;
    activeProtocols[0].balance = testValue;

    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
      />
    );

    selectProtocol(wrapper, 'input', 0);

    expect(wrapper.find(Overview).text()).toEqual(`Balance: ${testValue}`);
  });

  it('should render balance: 0, if balance is invalid', () => {
    activeProtocols[0].balance = -55;

    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
      />
    );

    selectProtocol(wrapper, 'input', 0);

    expect(wrapper.find(Overview).text()).toEqual(`Balance: 0`);
  });

  it('should render price equiv when an input protocol AND an output protocol are selected', () => {
    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
      />
    );

    expect(wrapper.find(Overview).length).toEqual(0);

    selectProtocol(wrapper, 'input', 0);
    expect(wrapper.find(Overview).length).toEqual(1);

    selectProtocol(wrapper, 'output', 0);
    expect(wrapper.find(Overview).length).toEqual(2);

    expect(wrapper.find(Overview).at(1).text().replace(/\s/g, '')).toEqual('1BTC=27.8666057561ETH');
  });

  it('should still render price equiv when no decimals are set', () => {
    const activeProtos = JSON.parse(JSON.stringify(activeProtocols)) as IProtocol[];
    activeProtos[0].decimals = undefined;
    activeProtos[1].decimals = undefined;

    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtos,
        }}
      />
    );

    expect(wrapper.find(Overview).length).toEqual(0);

    selectProtocol(wrapper, 'input', 0);
    expect(wrapper.find(Overview).length).toEqual(1);

    selectProtocol(wrapper, 'output', 0);
    expect(wrapper.find(Overview).length).toEqual(2);

    expect(wrapper.find(Overview).at(1).text().replace(/\s/g, '')).toEqual('1BTC=27.86660576ETH');
  });

  it('should not be possible to select the same protocol as input and output', () => {
    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
      />
    );

    const inputProto = selectProtocol(wrapper, 'input', 0).getDOMNode() as HTMLButtonElement;
    const outputProto = selectProtocol(wrapper, 'output', 0).getDOMNode() as HTMLButtonElement;

    expect(inputProto.title).not.toEqual(outputProto.title);
  });

  /**
   * SWAP FEATURE
   */

  it('should swap values accordingly, when clicking on the swap button', () => {
    const inputInitialVal = 12345;

    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
      />
    );

    selectProtocol(wrapper, 'input', 0);
    selectProtocol(wrapper, 'output', 0);

    const input = setInputValue(wrapper, 'input', inputInitialVal);
    wrapper.find(SwapButton).simulate('click');
    expect(input.prop('value')).not.toEqual(inputInitialVal);
  });

  /**
   * SUBMISSION
   */

  it('should submit if requirements are met', () => {
    activeProtocols[0].balance = 5;
    activeProtocols[1].balance = 10;

    const onSubmit = jest.fn((params: ISwapSubmission) => {
      expect(JSON.stringify(params)).toMatchSnapshot();
    });

    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
        onSubmit={onSubmit}
      />
    );

    selectProtocol(wrapper, 'input', 0);
    selectProtocol(wrapper, 'output', 0);
    setInputValue(wrapper, 'input', 4);

    wrapper.find(SUBMIT_BUTTON_SELECTOR).simulate('click');
    expect(onSubmit).toHaveBeenCalled();
  });

  it('should allow submission when conditions are met', () => {
    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
      />
    );

    selectProtocol(wrapper, 'input', 0);
    selectProtocol(wrapper, 'output', 0);

    // sets amount to 1, that is the minimum generated balance in activeProtocols mock
    setInputValue(wrapper, 'input', 1);

    expect(wrapper.find(SUBMIT_BUTTON_SELECTOR).text()).toEqual('Swap');
    expect(wrapper.find(SUBMIT_BUTTON_SELECTOR).getDOMNode()).toBeEnabled();
  });

  it('should prevent submission if locked prop is provided', () => {
    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
      />
    );

    expect(wrapper.find(SUBMIT_BUTTON_SELECTOR).getDOMNode()).toBeDisabled();
    expect(wrapper.find(SUBMIT_BUTTON_SELECTOR).text()).toEqual('Swap');
  });

  it("should prevent sumission if balance is insufficient, submit button's value should visually reflect that", () => {
    const wrapper = mount(
      <CurrencySwap
        maxFractionDigits={10}
        protocols={{
          input: activeProtocols,
        }}
      />
    );

    selectProtocol(wrapper, 'input', 0);
    selectProtocol(wrapper, 'input', 1);

    setInputValue(wrapper, 'input', 999999);

    expect(wrapper.find(SUBMIT_BUTTON_SELECTOR).getDOMNode()).toBeDisabled();
    expect(wrapper.find(SUBMIT_BUTTON_SELECTOR).text()).toEqual('Insufficient balance');
  });

  it('should not be possible to swap if locked prop is true', () => {
    const wrapper = mount(
      <CurrencySwap
        protocols={{
          input: activeProtocols,
        }}
        locked={true}
      />
    );

    selectProtocol(wrapper, 'input', 0);
    selectProtocol(wrapper, 'output', 0);

    // sets amount to 1, that is the minimum generated balance in activeProtocols mock
    setInputValue(wrapper, 'input', 1);

    expect(wrapper.find(SUBMIT_BUTTON_SELECTOR).text()).toEqual('Unlock Wallet');
    expect(wrapper.find(SUBMIT_BUTTON_SELECTOR).getDOMNode()).toBeDisabled();
  });

  it('should not render incoherent prive equivalence when input price is < 0', () => {
    activeProtocols[0].price = -1;

    const wrapper = mount(
      <CurrencySwap
        protocols={{
          input: activeProtocols,
        }}
        locked={true}
      />
    );

    selectProtocol(wrapper, 'input', 0);
    selectProtocol(wrapper, 'output', 0);

    expect(wrapper.find(Overview).at(2)).toEqual({});
  });

  /**
   * ACCESSIBILITY
   */

  it('should not have any accessibility issues', async () => {
    const { container } = render(
      <CurrencySwap
        protocols={{
          input: activeProtocols,
        }}
      />
    );

    expect(await axe(container)).toHaveNoViolations();
  });
});
