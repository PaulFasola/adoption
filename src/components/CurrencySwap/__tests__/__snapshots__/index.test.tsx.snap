// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencySwap component should submit if requirements are met 1`] = `"{\\"amounts\\":{\\"input\\":4,\\"output\\":111.46642302},\\"protocols\\":{\\"input\\":{\\"balance\\":5,\\"name\\":\\"Bitcoin\\",\\"symbol\\":\\"BTC\\",\\"decimals\\":8,\\"price\\":61000,\\"logoURI\\":\\"https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/btc.svg\\",\\"hidden\\":true},\\"output\\":{\\"balance\\":10,\\"name\\":\\"Ethereum\\",\\"symbol\\":\\"ETH\\",\\"decimals\\":18,\\"price\\":2189,\\"logoURI\\":\\"https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/eth.svg\\",\\"hidden\\":true}}}"`;
